README
README for Perl extension WWW::Curl

The perl module WWW::Curl provides an interface to the cURL library "libcurl".
See http://curl.haxx.se/ for more information on cURL and libcurl.

This module requires libcurl and the corresponding headerfiles to be
installed. You then may install this module via the usual way
(for installation on Windows please see README.Win32):

    perl Makefile.PL
    make
    make test
    make install

The Makefile.PL script will try to use the 'curl-config' command to find
the correct headers and libcurl path. If curl.h can't be found,
either through curl-config or the internal list of guesses, you can specify
an explicit path when building the Makefile:

    perl Makefile.PL /home/foo/curl/include

Minimum version requirements:

This module, starting from version 4 requires at least Perl 5.6 and libcurl 7.10.8.
These releases are more than 5 years old and they have multiple security vulnerabilities,
it is advised that you upgrade as soon as possible to a more recent, secure version.
Anything older than these versions of Perl and libcurl respectively are not supported.

Windows specific fixes and patches are welcome as testing is mainly focused on linux.

The module provides the same functionality, except as noted in the documentation, 
as libcurl provides to C programs. Please refer to the documentation of libcurl for the
general interface description and the WWW::Curl POD for the differences.

This software is distributed AS IS, WITHOUT WARRANTY OF ANY KIND, either
express or implied.

History:

The author of the original relase of this software is Georg Horn <horn at koblenz-net.de>

Parts of the callback support were added Forrest Cahoon
<forrest.cahoon at merrillcorp.com>

More callback support, many tests additional documentation and Makefile
features have been added by Cris Bailiff <c.bailiff+curl at devsecure.com>

Curl multi support has been added by Sebastian Riedel <sri at oook.de>

The current maintainer is Cris Bailiff <c.bailiff+curl at devsecure.com>

The latest version can be downloaded from http://curl.haxx.se/libcurl/perl/ or
found on CPAN as the module WWW::Curl
 
Copyright (C) 2000-2005, 2008 Daniel Stenberg, Cris Bailiff , et al.  
You may opt to use, copy, modify, merge, publish, distribute and/or sell
copies of the Software, and permit persons to whom the Software is
furnished to do so, under the terms of the MPL or the MIT/X-derivate 
licenses. You may pick one of these licenses.


Changes
Revision history for Perl extension WWW::Curl.
4.06 Sun Apr 05 2009: - Balint Szilakszi <szbalint at cpan.org>

    - Fixed a setopt issue that could cause a setopt string to get
      corrupted with pre 7.17.0 libcurl versions.
      (Thanks to Andreas Marienborg for the excellent bugreport and testcase!)
    - Made the Makefile.PL curl.h parsing slightly more generic to detect a
      wider selection of constants (fixes debian bugs #519841 and #518942).
      This makes it possible to use CURLE_* and CURL_NETRC_* constants.
    - Added a clearly invalid protocol to t/18twinhandles.t (fixes RT #41425).

4.05 Sat Jul 06 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Corrected the duphandle mechanism.
    - Fixed a typo (fixes RT #37253).
    - Added better Windows compatibility detection and separate
      installation instructions (thanks to Shiyu Tang).

4.04 Sat Jun 21 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Test::More and use_ok at begin time doesn't mesh with no plan.

4.03 Sat Jun 21 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Removing option to pass destination files as a typeglob as
      it is a backwards pointing magic and caused bugs.
      Use a filehandle or reference instead.
    - Fixed tests for 5.6, raised minimum Perl version to 5.6.1 due to tests.
    - Fixed a duphandle bug.

4.02 Thu Jun 19 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Adding forgotten backwards compatibility with older Perl versions

4.01 Wed Jun 18 2008: - Balint Szilakszi <szbalint at cpan.org>

    - Reimplemented callback handling to fix several I/O issues (fixes RT #35491)
    - Documentation showing example WWW::Curl::Multi usage is corrected (fixes RT #35520)
    - Should now compile on HP-UX and OSX (fixes RT #36058 and #35349)
    - POD validity fix (fixes RT #15850)
    - Added several new tests
    - Upgraded to Module::Install 0.75 as it now exits correctly on build time
      with NA if the required libcurl isn't available.

4.00 Wed Mar 05 2008: - Balint Szilakszi <szbalint at cpan.org>
    - Rewritten documentation.
    - New test suite.
    - New build system: using Module::Install instead of ExtUtils::Makemaker.
    - New minimal requirements, Perl 5.6 and libcurl 7.10.8.
    - New WWW::Curl::Multi interface.
    - Removed WWW::Curl::easy.
    - Removed compatibility code for pre 7.10.8 libcurl.
    - Added lots of small improvements to setopt and friends.
    - Bugfixes, mainly for memory leaks.
    - SIGNATURE.

3.12 Wed Feb 27 2008: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - New Share.pm.in was missed from the MANIFEST, so was not in the distribution. Fixed.

3.11 Mon Feb 25 2008: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Merge recent contributions.
    - Updates from various small bug/build reports

3.10 Fri Feb 22 2008: - Anton Fedorov <datacompboy at mail.ru>
    - Fix documentation for multi interface
    - Added WWW::Curl::Share
    - Added strerror into Easy and Multi interfaces
    - Added support for CURLOPT_DEBUGFUNCTION in Easy interface

3.02_01 Tue Jan 29 2008: - Mark Hindley
    - Contributed CURLOPT_*_LARGE support

3.02_00 Fri Sep 16 2005: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Workaround case-sensitive filename issues by making equivalent Easy
      names in easy namespace
    - Test scripts - Fixup ssl tests and remove leftover MUTE option
    - NOTE: That due to namespace changes (easy->Easy) you must change any existing
      code to 'use WWW::Curl::Easy' instead of 'use WWW::Curl::easy', but
      you can still use the WWW::Curl::easy function names until they
      are removed in a future release.
    - Tested on Redhat EL3 (curl 7.10.6) and Mandrake 10.1 (curl 7.13.1)

3.01  Thu Apr 20 2004: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Merged Sebastians changes into CVS
    - Updated remaining tests to Test::Simple
    - Added backwards compatability to V2 names (WWW::Curl::easy etc)
    - Dropped backwards compatability to V1.x names

3.00  Thu Feb 12 2004: - Sebastian Riedel <sri at oook.de>
    - Added multi support
    - New module names
    - New tests
    - New documentation
    - New examples
    - Big cleanup!!!
    - Too much to tell, it's worth an upgrade

2.00  Tue Apr 22 2003: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - New top level package name of WWW::Curl in preparation for
      entry to CPAN
    - Rename "Curl::easy" to "WWW::Curl::easy"
    - Add backwards compatability namespace module for existing scripts
    - Implement initial curl_easy_duphandle support
    - Started on curl_easy_form support (WWW:Curl::form) - NOT FUNCTIONAL YET
    - Fixup use of env vars in t/07ftp-upload.t (jellyfish at pisem.net)
    - Adjust IP addresses for t/08ssl.t tests due to moved https servers

1.35  Sun Sep 22 2002: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Fixed progress function callback prototype [ curl-Bugs-612432 ],
      reflecting the fix made in curl-7.9.5. Tested in t/05progress.t to
      now return sensible values!

1.34  Wed Aug  7 2002: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Fix off-by-one error in setting up of curl_slists from perl arrays,
      which was causing the last item of slists to be dropped. Added regression
      test case.

1.33  Mon Aug  5 2002: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Fix serious bug in read callback support (used for POST and upload
      requests), introduced in 1.30, which uploaded random data (due to a
      reversed src/dest in a memory copy).

1.32  Thu Aug  1 2002: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Minor Makefile.PL fixes to build cleanly with curl 7.8 as found
      on redhat 7.2.

1.31  Tue Jul 16 2002: - Cris Bailiff <c.bailiff+curl at devsecure.com>
    - Generate better switch() statement syntax in C code, to fix build
      issues on some systems with strict compilers. Reported by Ignasi Roca.


Makefile.PL
# Makefile.PL for Perl module WWW::Curl
# Check out the README file for more information.

use inc::Module::Install;

name			'WWW-Curl';
abstract		'Perl extension interface for libcurl';
author			'Cris Bailiff <c.bailiff+curl at devsecure.com>';
license			'MPL or MIT/X-derivate';
perl_version		'5.006001';
no_index		directory => 'template';
# This is a hack. If you have libcurl installed, just specify curl.h below
# and comment out this line.
if ($^O ne 'MSWin32') {
	requires_external_bin	'curl-config';
} else {
	print	"Sorry, no automated install is available on Windows,\n".
		"please see the README.Win32 file on instructions for a manual install.\n";
	exit(0);
}

my @includes = qw();
my ($cflags,$lflags, $ldflags) = ('','','');

# This utility helper generates the constants function from curl.h
# It is normally only used by the maintainer, but if you're curl is older
# or missing some constants, you can delete curlopt-constants.c and re-run 'perl Makefile.PL'
# You may need to specify where to find curl.h on your platform
# These are guesses only, in case curl-config is not telling us.

if ($^O ne 'MSWin32') {
	push @includes, qw(
		/usr/include
		/usr/local/curl/include
		/usr/local/include/curl
		../../include
		../curl/include
	);
}
#
# Get curl to tell us where it is, if we can.
#


if ($^O ne 'MSWin32') {
	$cflags = `curl-config --cflags`;
	$lflags = `curl-config --libs`;
}

# can't find link flags, make some guesses
if (!defined($lflags)) {
    $lflags="-lcurl";
    print "Guessing your linker flags as: $lflags\n";
}

my ($flag) = ($cflags =~ m/-I(\S+)/);

if (defined $flag) {
    unshift @includes, $flag;    # first guess
}

# try the path given on the command line, if any
if (defined($ARGV[0])) {
    unshift @includes, $ARGV[0];
};

my $curl_d = "";
my $curl_h;

# otherwise try a list of common locations
foreach my $try (@includes) {
    if (-f $try . "/curl/curl.h") {
        $curl_d = $try;
        $curl_h = $try . "/curl/curl.h";
        last;
    }
}

if (!defined($curl_h)) {
    die "Cannot find curl.h - cannot build constants files  - see Makefile.PL";
} else {
    $curl_d = "-I" . $curl_d;
    print "Found curl.h in $curl_h\n";
    open(CURL_H, "<" . $curl_h) or die "Can't open curl.h\n";
    my %constants;
    while (<CURL_H>) {
        if ($_ =~ m/CINIT\(/ and $_ !~ m/#/) {
            my ($option, $type, $code) =
                m/.*CINIT\((\w*)\s*,\s*(\w+)\s*,\s*(\d+).*/;
	    $constants{CURLOPT_}->{$option} = $option;
	} elsif ($_ =~ m/^#define CURLOPT_\w+\s+CURLOPT_\w+/) {
            my ($option, $value) =
                m/^#define CURLOPT_(\w+)\s+CURLOPT_(\w+)/;
	    $constants{CURLOPT_}->{$option} = $value;
        } elsif ($_ =~ m/^\s*(CURLINFO_|CURLSHOPT_|CURL_LOCK_DATA_|CURLE_|CURL_NETRC_)(\w+)/) {
	    $constants{$1}->{$2} = $2;
        }

    }
    close(CURL_H);

    # some things are ifdefed out...
    foreach my $ifdef0 (qw(FLAGS PROGRESSMODE MOREDOCS)) {
        delete $constants{CURLOPT_}->{$ifdef0};
    }

    print "Building curlopt-constants.c for your libcurl version\n";

    open(CURL_XS, ">curlopt-constants.c")
        or die "Can't write curlopt-constants.c\n";

    # boilerplate xs constant function here
    print CURL_XS <<HERE
static int
constant(char *name, int arg)
{
    errno = 0;
HERE
        ;
for my $group (sort keys %constants) {
	my $grouplength = length($group);
	my $groupref = $constants{$group};
	my @constants = keys %{$constants{$group}};
	print CURL_XS <<HERE2

    if (strncmp(name, "$group", $grouplength) == 0) {
        name += $grouplength;
        switch (*name) {
HERE2
        ;

    for my $next_initial ('A' .. 'Z') {
        print CURL_XS "        case '$next_initial':\n";
        my $count = 0;
        foreach my $option (sort @constants) {
            my $initial = substr($option, 0, 1);
            if ($next_initial eq $initial) {

                print CURL_XS
"            if (strEQ(name, \"$option\")) return $group"."$groupref->{$option};\n";

                $count++;
            }
        }
        if ($count or $next_initial eq 'Z') {
            print CURL_XS "            break;\n";
        }
    }

    print CURL_XS "    };\n";
    print CURL_XS "    }\n";
}

    print CURL_XS <<HERE

    errno = EINVAL;
    return 0;
}
HERE
        ;

    close(CURL_XS);

    print "Building Easy.pm constants for your libcurl version\n";

    open(EASY_PM,    ">lib/WWW/Curl/Easy.pm")    or die "Can't create lib/WWW/Curl/Easy.pm\n";
    open(EASY_PM_IN, "template/Easy.pm.tmpl") or die "Can't read template/Easy.pm.tmpl\n";
    while (my $line = <EASY_PM_IN>) {
        if ($line !~ m/^\@CURLOPT_INCLUDE\@/) {
			print EASY_PM $line;
        } else {
	for my $group (qw/CURLOPT_ CURLINFO_ CURLE_ CURL_NETRC_/) {
            for my $option (sort keys %{$constants{$group}}) {
                next unless $option;
                print EASY_PM $group.$option."\n";
            }
        }
	}
    }
    close(EASY_PM);
    close(EASY_PM_IN);

	print "Building Share.pm constants for your libcurl version\n";

    open(SHARE_PM,   ">lib/WWW/Curl/Share.pm")    or die "Can't create lib/WWW/Curl/Share.pm\n";
    open(SHARE_PM_IN, "template/Share.pm.tmpl") or die "Can't read template/Share.pm.tmpl\n";
    while (my $line = <SHARE_PM_IN>) {
        if ($line !~ m/^(.*?)\@CURLSHOPT_INCLUDE\@/) {
            print SHARE_PM $line;
        } else {
            foreach my $option (sort keys %{$constants{CURLSHOPT_}}) {
                print SHARE_PM $1 . "CURLSHOPT_".$option . "\n";
            }
            foreach my $option (sort keys %{$constants{CURL_LOCK_DATA_}}) {
                print SHARE_PM $1 . "CURL_LOCK_DATA_".$option . "\n";
            }
        }
    }
    close(SHARE_PM);
    close(SHARE_PM_IN);
}

# Let Module::Install generate META.yml and other necessary files.
WriteMakefile(
    'NAME'         => 'WWW::Curl',
    'VERSION_FROM' => 'lib/WWW/Curl.pm',      # finds $VERSION
    'LIBS'         => "$ldflags $lflags",        # e.g., '-lm'
    'INC'          => $curl_d,        # e.g., '-I/usr/include/other'
    'clean' => { FILES => "curlopt-constants.c head.out body.out" }
);



lib/WWW/Curl/Share.pm
package WWW::Curl::Share;

use strict;
use Carp;
use vars qw(@ISA @EXPORT @EXPORT_OK $AUTOLOAD);

use WWW::Curl;
require Exporter;
require AutoLoader;

@ISA = qw(Exporter DynaLoader);

@EXPORT = qw(
CURLSHOPT_LAST
CURLSHOPT_LOCKFUNC
CURLSHOPT_NONE
CURLSHOPT_SHARE
CURLSHOPT_UNLOCKFUNC
CURLSHOPT_UNSHARE
CURLSHOPT_USERDATA
CURL_LOCK_DATA_CONNECT
CURL_LOCK_DATA_COOKIE
CURL_LOCK_DATA_DNS
CURL_LOCK_DATA_LAST
CURL_LOCK_DATA_NONE
CURL_LOCK_DATA_SHARE
CURL_LOCK_DATA_SSL_SESSION
);

sub AUTOLOAD {

    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    ( my $constname = $AUTOLOAD ) =~ s/.*:://;
    return constant( $constname, 0 );
}

1;
__END__

=head1 NAME

WWW::Curl::Share - Perl extension interface for libcurl

=head1 SYNOPSIS

	use WWW::CURL::Share;
	my $curlsh = new WWW::Curl::Share;
	$curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_COOKIE);
	$curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
	$curl->setopt(CURLOPT_SHARE, $curlsh);
	$curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_COOKIE);
	$curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_DNS);

=head1 DESCRIPTION

WWW::Curl::Share is an extension to WWW::Curl::Easy
which makes it possible to use single cookies/dns cache for
several Easy handles.

=head1 METHODS

	$curlsh = new WWW::Curl::Share
		This method constructs a new WWW::Curl::Share object.

	$curlsh->setopt(CURLSHOPT_SHARE, $value );
		Enables share for:
			CURL_LOCK_DATA_COOKIE	use single cookies database
			CURL_LOCK_DATA_DNS	use single DNS cache
	$curlsh->setopt(CURLSHOPT_UNSHARE, $value );
		Disable share for given $value (see CURLSHOPT_SHARE)

	$curlsh->strerror( ErrNo )
		This method returns a string describing the CURLSHcode error 
		code passed in the argument errornum.

	$curl->setopt(CURLOPT_SHARE, $curlsh)
		Attach share object to WWW::Curl::Easy instance

	List of all available options and lock constants:
		CURLSHOPT_LAST
		CURLSHOPT_LOCKFUNC
		CURLSHOPT_NONE
		CURLSHOPT_SHARE
		CURLSHOPT_UNLOCKFUNC
		CURLSHOPT_UNSHARE
		CURLSHOPT_USERDATA
		CURL_LOCK_DATA_CONNECT
		CURL_LOCK_DATA_COOKIE
		CURL_LOCK_DATA_DNS
		CURL_LOCK_DATA_LAST
		CURL_LOCK_DATA_NONE
		CURL_LOCK_DATA_SHARE
		CURL_LOCK_DATA_SSL_SESSION

=head1 AUTHOR

Anton Fedorov (datacompboy <at> mail.ru)

=head1 COPYRIGHT

Copyright (C) 2004 Sebastian Riedel, et al.

You may opt to use, copy, modify, merge, publish, distribute and/or sell
copies of the Software, and permit persons to whom the Software is furnished
to do so, under the terms of the MPL or the MIT/X-derivate licenses. You may
pick one of these licenses.

=head1 SEE ALSO

WWW::Curl, WWW::Curl::Easy, http://curl.haxx.se/


lib/WWW/Curl.pm
package WWW::Curl;

use strict;
use warnings;
use vars qw(@ISA $VERSION);
use DynaLoader;

BEGIN {
    $VERSION = '4.06';
    @ISA     = qw(DynaLoader);
    __PACKAGE__->bootstrap;
}

1;

__END__

=head1 NAME

WWW::Curl - Perl extension interface for libcurl

=head1 SYNOPSIS

    use WWW::Curl;
    print $WWW::Curl::VERSION;


=head1 DESCRIPTION

WWW::Curl is a Perl extension interface for libcurl.

=head1 DOCUMENTATION

This module provides a Perl interface to libcurl. It is not intended to be a standalone module
and because of this, the main libcurl documentation should be consulted for API details at
L<http://curl.haxx.se>. The documentation you're reading right now only contains the Perl specific
details, some sample code and the differences between the C API and the Perl one.

=head1 WWW::Curl::Easy

The name might be confusing, it originates from libcurl. This is not an ::Easy module
in the sense normally used on CPAN.

Here is a small snippet of making a request with WWW::Curl::Easy.

	use strict;
	use warnings;
	use WWW::Curl::Easy;

	# Setting the options
	my $curl = new WWW::Curl::Easy;
	
	$curl->setopt(CURLOPT_HEADER,1);
	$curl->setopt(CURLOPT_URL, 'http://example.com');
	my $response_body;

	# NOTE - do not use a typeglob here. A reference to a typeglob is okay though.
	open (my $fileb, ">", \$response_body);
	$curl->setopt(CURLOPT_WRITEDATA,$fileb);

	# Starts the actual request
	my $retcode = $curl->perform;

	# Looking at the results...
	if ($retcode == 0) {
		print("Transfer went ok\n");
		my $response_code = $curl->getinfo(CURLINFO_HTTP_CODE);
		# judge result and next action based on $response_code
		print("Received response: $response_body\n");
	} else {
		print("An error happened: ".$curl->strerror($retcode)." ($retcode)\n");
	}


=head1 WWW::Curl::Multi

	use strict;
	use warnings;
	use WWW::Curl::Easy;
	use WWW::Curl::Multi;

	my %easy;
	my $curl = WWW::Curl::Easy->new;
	my $curl_id = '13'; # This should be a handle unique id.
	$easy{$curl_id} = $curl;
	my $active_handles = 0;

	$curl->setopt(CURLOPT_PRIVATE,$curl_id);
	# do the usual configuration on the handle
	...

	my $curlm = WWW::Curl::Multi->new;
	
	# Add some easy handles
	$curlm->add_handle($curl);
	$active_handles++;

	while ($active_handles) {
		my $active_transfers = $curlm->perform;
		if ($active_transfers != $active_handles) {
			while (my ($id,$return_value) = $curlm->info_read) {
				if ($id) {
					$active_handles--;
					my $actual_easy_handle = $easy{$id};
					# do the usual result/error checking routine here
					...
					# letting the curl handle get garbage collected, or we leak memory.
					delete $easy{$id};
				}
			}
		}
	}

This interface is different than what the C API does. $curlm->perform is non-blocking and performs
requests in parallel. The method does a little work and then returns control, therefor it has to be called
periodically to get the job done. It's return value is the number of unfinished requests.

When the number of unfinished requests changes compared to the number of active handles, $curlm->info_read
should be checked for finished requests. It returns one handle and it's return value at a time, or an empty list
if there are no more finished requests. $curlm->info_read calls remove_handle on the given easy handle automatically,
internally. The easy handle will still remain available until it goes out of scope, this action just detaches it from
multi.

Please make sure that the easy handle does not get garbage collected until after the multi handle finishes processing it,
or bad things happen.

The multi handle does not need to be cleaned up, when it goes out of scope it calls the required cleanup methods
automatically.

It is possible to use $curlm->add_handle to add further requests to be processed after $curlm->perform has been called.
WWW::Curl::Multi doesn't care about the order. It is possible to process all requests for a multi handle and then add
a new batch of easy handles for processing.

=head1 WWW::Curl::Share

	use WWW::CURL::Share;
	my $curlsh = new WWW::Curl::Share;
	$curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_COOKIE);
	$curlsh->setopt(CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
	$curl->setopt(CURLOPT_SHARE, $curlsh);
	$curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_COOKIE);
	$curlsh->setopt(CURLSHOPT_UNSHARE, CURL_LOCK_DATA_DNS);

WWW::Curl::Share is an extension to WWW::Curl::Easy which makes it possible
to use a single cookies/dns cache for several Easy handles.

It's usable methods are:
	
	$curlsh = new WWW::Curl::Share
		This method constructs a new WWW::Curl::Share object.

	$curlsh->setopt(CURLSHOPT_SHARE, $value );
		Enables share for:
			CURL_LOCK_DATA_COOKIE	use single cookies database
			CURL_LOCK_DATA_DNS	use single DNS cache
	$curlsh->setopt(CURLSHOPT_UNSHARE, $value );
		Disable share for given $value (see CURLSHOPT_SHARE)

	$curlsh->strerror( ErrNo )
		This method returns a string describing the CURLSHcode error 
		code passed in the argument errornum.

This is how you enable sharing for a specific WWW::Curl::Easy handle:

	$curl->setopt(CURLOPT_SHARE, $curlsh)
		Attach share object to WWW::Curl::Easy instance


=head1 COMPATIBILITY

=over

=item curl_easy_setopt

Most of the options should work, however some might not. Please send reports, tests and patches to fix
those. 

=item curl_easy_escape

Not implemented. Since equivalent Perl code is easily produced, this method will only made
available for interface completeness, if ever.

=item curl_easy_init

Used only internally. The standard Perl way of initializing an object should be used,
 C<< my $curl = WWW::Curl::Easy->new; >>.

=item curl_easy_cleanup

Used only internally. Curl object cleanup happens when the handle goes out of scope.

=item curl_easy_duphandle

Should be working for most cases, however do not change the value of options which accept
a list/arrayref value on a duped handle, otherwise memory leaks or crashes will happen.
This behaviour will be fixed in the future.

=item curl_easy_pause

Not implemented.

=item curl_easy_reset

Not implemented.

=item curl_easy_unescape

Not implemented. Trivial Perl replacements are available.

=item curl_escape

Not implemented and won't be as this method is considered deprecated.

=item curl_formadd

Not yet implemented.

=item curl_formfree

When WWW::Curl::Form support is added, this function will be used internally,
but won't be accessible from the public API.

=item curl_free

Used internally. Not exposed through the public API, as this call has no relevance
to Perl code.

=item curl_getdate

Not implemented. This function is easily replaced by Perl code and as such, most likely
it won't be implemented.

=item curl_global_cleanup

Only used internally, not exposed through the public API.

=item curl_global_init

Only used internally, not exposed through the public API.

=item curl_global_init_mem

Not implemented.

=item curl_slist_append

Only used internally, not exposed through the public API.

=item curl_slist_free_all

Only used internally, not exposed through the public API.

=item curl_unescape

Not implemented and won't be, as this method is considered deprecated.

=item curl_version_info

Not yet implemented.

=item curl_multi_*

Most methods are either not exposed through the WWW::Curl::Multi API or they behave differently
than it's C counterpart. Please see the section about WWW::Curl::Multi above.

=back

=head1 USAGE CASES

The standard Perl WWW module, LWP should be used in most cases to work with
HTTP or FTP from Perl. However, there are some cases where LWP doesn't
perform well. One is speed and the other is parallelism. WWW::Curl is much faster,
uses much less CPU cycles and it's capable of non-blocking parallel requests.

In some cases, for example when building a web crawler, cpu usage and parallel downloads are
important considerations. It can be desirable to use WWW::Curl to do the heavy-lifting of
a large number of downloads and wrap the resulting data into a Perl-friendly structure by
HTTP::Response.

=head1 CHANGES

Version 4.01 - 4.06 adds several bugfixes and extends functionality coverage. See Changes file.

Version 4.00 added new documentation, the build system changed to Module::Install,
the test suite was rewritten to use Test::More, a new calling syntax for WWW::Curl::Multi
was added, memory leak and other bugfixes added, Perl 5.6 and libcurl 7.10.8 as minimum
requirements for this module were set.

Version 3.12 is a bugfix for a missing Share.pm.in file in the release.

Version 3.11 added WWW::Curl::Share.

Version 3.10 adds the WWW::Curl::Share interface by Anton Federov
and large file options after a contribution from Mark Hindley.

Version 3.02 adds some backwards compatibility for scripts still using
'WWW::Curl::easy' names.

Version 3.01 added some support for pre-multi versions of libcurl.

Version 3.00 adds WWW::Curl::Multi interface, and a new module names
following perl conventions (WWW::Curl::Easy rather than WWW::Curl::easy),
by Sebastian Riedel <sri at cpan.org>.

Version 2.00 of WWW::Curl::easy is a renaming of the previous version
(named Curl::easy), to follow CPAN naming guidelines, by Cris Bailiff.

Versions 1.30, a (hopefully) threadable, object-oriented,
multiple-callback compatible version of Curl::easy was substantially
reworked from the previous Curl::easy release (1.21) by Cris Bailiff.

=head1 AUTHORS

Currently maintained by Cris Bailiff <c.bailiff+curl at devsecure.com>

Original Author Georg Horn <horn@koblenz-net.de>, with additional callback,
pod and test work by Cris Bailiff <c.bailiff+curl@devsecure.com> and
Forrest Cahoon <forrest.cahoon@merrillcorp.com>. Sebastian Riedel added ::Multi
and Anton Fedorov (datacompboy <at> mail.ru) added ::Share. Balint Szilakszi
repackaged the module into a more modern form.

=head1 COPYRIGHT

Copyright (C) 2000-2005,2008 Daniel Stenberg, Cris Bailiff,
Sebastian Riedel, Balint Szilakszi et al.

You may opt to use, copy, modify, merge, publish, distribute and/or sell
copies of the Software, and permit persons to whom the Software is furnished
to do so, under the terms of the MPL or the MIT/X-derivate licenses. You may
pick one of these licenses.

=head1 SEE ALSO

http://curl.haxx.se
